### DISCUSSION TITLE
Lazy imports

### DISCUSSION THREAD

**Discussion by gryznar** [modular/modular#322] [Q&A] [answered]
Will Mojo have lazy imports by default? With only pointing to module in uninitialized state without memory allocation and module execution?

**Comment by mojodojodev** [comment]
I've tested this and currently it doesn't, however Python 3.12 adds it: https://peps.python.org/pep-0690/

And also adds it to the C-API if you scroll down the page there, which Mojo will then be able to use

**Comment by gryznar** [comment]
Steering Council rejected this PEP to not differentiate community

**Reply by mojodojodev** [reply]
Oh yeah, here's the steering council response for anyone interested:

>We agree with the widely accepted sentiment that faster Python startup time is desirable. Large command line tools in particular suffer as that is a human user experience. Lazy imports, as proposed, are one of many potential mechanisms that can help with that.
>
>But a problem we deem significant when adding lazy imports as a language feature is that it becomes a split in the community over how imports work. A need to test code both ways in both traditional and lazy import setups arises. It creates a divergence between projects who expect and rely upon import time code execution and those who forbid it. It also introduces the possibility of unexpected import related exceptions occurring in code at the time of first use virtually anywhere. Such exceptions could bubble up from transitive dependency first use in unanticipated places.
>
>A world in which Python only supported imports behaving in a lazy manner would likely be great. But we cannot rewrite history and make that happen. As we do not envision the Python langauge transitioning to a world where lazy imports are the default, let alone only, import behavior. Thus introducing this concept would add complexity to our ecosystem.
>
>We also discussed specific implementation details of this PEP that many of us did not really like, but ultimately decided that even with mere internal implementation changes it didn’t change our broader feeling on the matter: The complexity of fragmentation was bad and we were going to say “no” regardless.

**Comment by ksandvik** [comment]
It might be that lazy importing might even be an optimization that the developers would not need to worry about.

**Comment by czheo** [comment, marked as answer]
Based on my understanding, Mojo is a compiled language. Statically linked lib will just go into the executable, so will be loaded into memory during startup. Dynamically linked lib will be `dlopen`ed during runtime, so is "lazy".
